### PR TITLE
Relationship generate

### LINKED ISSUE #12187: Support generating realtionships using the mesheryctl generate command.
Source relationships from spreadsheet itself while generating models generate the respective relationships.

### PR DESCRIPTION
**Pull request by Jougan-0**
**Notes for Reviewers**

This PR fixes #12187 #12186

Need to clean the directory remove redundant models and move current relationships as an archive.
**[Signed commits](https://github.com/meshery/meshery/blob/master/CONTRIBUTING.md#signing-off-on-commits-developer-certificate-of-origin)**
- [x] Yes, I signed my commits.

### CODE REVIEW

File: mesheryctl/internal/cli/root/registry/generate.go [hunk 1]
```diff
@@ -336,13 +354,24 @@ func InvokeGenerationFromSheet(wg *sync.WaitGroup) error {
...
+	utils.ProcessRelationships(relationshipCSVHelper)
 	wgForSpreadsheetUpdate.Wait()
 	return nil
 }
 
 // For registrants eg: meshery, whose components needs to be directly created by referencing meshery/schemas repo.
 // the sourceURL contains the path of models component definitions
 func GenerateDefsForCoreRegistrant(model utils.ModelCSV, ComponentCSVHelper *utils.ComponentCSVHelper) error {
+	registrant := "meshery"
+	if models, exists := ComponentCSVHelper.Components[registrant]; exists {
```
Comment (MUzairS15) [inline mesheryctl/internal/cli/root/registry/generate.go:366]:
```suggestion
	if models, exists := ComponentCSVHelper.Components[registrant][model.Model]; exists {
```

And remove the if branch.

File: mesheryctl/internal/cli/root/registry/generate.go [hunk 2]
```diff
@@ -469,6 +498,18 @@ func parseComponentSheet(url string) (*utils.ComponentCSVHelper, error) {
...
 }
 
+func parseRelationshipSheet(url string) (*utils.RelationshipCSVHelper, error) {
+	relationshipCSVHelper, err := utils.NewRelationshipCSVHelper(url, "Relationships", relationshipSpredsheetGID, relationshipCSVFilePath)
+	if err != nil {
+		return nil, err
+	}
+	err = relationshipCSVHelper.ParseRelationshipsSheet(modelName)
+	if err != nil {
+		return nil, ErrGenerateModel(err, "unable to start model generation")
```
Comment (MUzairS15) [inline mesheryctl/internal/cli/root/registry/generate.go:508]: Update/define error according to operation.

File: mesheryctl/pkg/utils/relationship.go [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+	CSVPath        string
+	Relationships  []RelationshipCSV
+}
+type RelationshipCSV struct {
+	Model       string `json:"Model" csv:"Model"`
+	Version     string `json:"Version" csv:"Version"`
+	KIND        string `json:"kind" csv:"kind"`
+	Type        string `json:"type" csv:"type"`
+	SubType     string `json:"subType" csv:"subType"`
+	Description string `json:"metadata.description" csv:"metadata.description	"`
```
Comment (MUzairS15) [inline mesheryctl/pkg/utils/relationship.go:30]: Did you verified this tag? Does it work correctly? 
We might need an custom implementation for Unmarshal. (for xml it works, I believe but not for JSON).

File: mesheryctl/pkg/utils/relationship.go [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+
+	var fromKind, toKind string
+
+	if rel.Selectors != nil && len(*rel.Selectors) > 0 {
+		firstSelector := (*rel.Selectors)[0]
+		if len(firstSelector.Allow.From) > 0 && len(firstSelector.Allow.To) > 0 {
+			if firstSelector.Allow.From[0].Kind != nil {
+				fromKind = *firstSelector.Allow.From[0].Kind
+			}
+			if firstSelector.Allow.To[0].Kind != nil {
```
Comment (MUzairS15) [inline mesheryctl/pkg/utils/relationship.go:200]: This means there's a defined contract that in the sheet the relaationship definition will always have single selector, but why this constraint?

File: mesheryctl/pkg/utils/relationship.go [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+				toKind = *firstSelector.Allow.To[0].Kind
+			}
+		}
+	}
+	// Construct the filename
+	//filename:- kind-type-subtype-from-{selector from kind}-to-{selector to kind}.json
+	filename := fmt.Sprintf("%s-%s-%s-from-%s-to-%s.json",
```
Comment (MUzairS15) [inline mesheryctl/pkg/utils/relationship.go:207]: If the constraint above is for giving file a unique name, add a column in the sheet for "name".

File: mesheryctl/pkg/utils/relationship.go [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+		rel.Kind, rel.RelationshipType, rel.SubType,
+		utils.ReplaceSpacesAndConvertToLowercase(fromKind),
```
Comment (MUzairS15) [inline mesheryctl/pkg/utils/relationship.go:209]: Any specific reason? These kinds maps directly to "component.component.kind", and all of the consumers of  relationships performs the exact match.

If we are changing it ensure components defs and all other places are also updated to reflect this change.

**Review by MUzairS15 — CHANGES_REQUESTED** [review]

File: mesheryctl/pkg/utils/relationship.go (hunk 5 above)
Comment (Jougan-0) [reply mesheryctl/pkg/utils/relationship.go:207]: Okay

File: mesheryctl/pkg/utils/relationship.go (hunk 6 above)
Comment (Jougan-0) [reply mesheryctl/pkg/utils/relationship.go:209]: It was just for naming convection when was naming the files after splitting lee mentioned to use lowercase.

File: mesheryctl/pkg/utils/relationship.go (hunk 3 above)
Comment (Jougan-0) [reply mesheryctl/pkg/utils/relationship.go:30]: It did I'll recheck and confirm.

File: mesheryctl/internal/cli/root/registry/generate.go (hunk 1 above)
Comment (Jougan-0) [reply mesheryctl/internal/cli/root/registry/generate.go:366]: Okay

**Pull request merged**